### DISCUSSION TITLE
Building for M1 Macs

### DISCUSSION THREAD

**Discussion by PurpleBooth** [Homebrew/discussions#3571] [Tap maintenance and Homebrew development] [answered]
Is there a way to bottle for M1's on Github actions? I would love to start providing precompiled M1 bottles, but I can't find any examples for it. Is it possible?

I am already happily using the rather good workflow that is in the "tap-new" template.

**Comment by carlocab** [comment, marked as answer]
The only supported way of doing this is through the use of self-hosted M1 runners, which means it's not quite possible with GitHub-provides runners at the moment.

You could maybe get it to work with a lot of painful hacking at `brew`, but that's neither supported nor recommended.

**Reply by DeveloperKev** [reply]
> The only supported way of doing this is through the use of self-hosted M1 runners, which means it's not quite possible with GitHub-provides runners at the moment.
> 
> You could maybe get it to work with a lot of painful hacking at `brew`, but that's neither supported nor recommended.

how to install brew packages without installing homebrew?

**Reply by SMillerDev** [reply]
You can't

**Reply by DeveloperKev** [reply]
what is `cellar` directory? can i change it from `/opt/homebrew/Cellar` to local? 

Another thing, https://github.com/wbond/package_control/issues/1612
brew is conflicting with the crypto library, pls resolve it.

**Reply by SMillerDev** [reply]
> brew is conflicting with the crypto library, pls resolve it.

That's not what's in that issue

> what is cellar directory?

https://docs.brew.sh/Formula-Cookbook#homebrew-terminology

> can i change it from /opt/homebrew/Cellar to local?

I don't know what `local` is supposed to be in this context, but if you change the cellar location all your Homebrew software is going to break so this is a very bad idea.

**Reply by DeveloperKev** [reply]
i see the formula. there are executable files,
so how do i do the installation manually?

**Reply by SMillerDev** [reply]
That depends entirely on the software you want to install and the state of your system. For example for OpenSSL the install instructions are here: https://github.com/openssl/openssl#build-and-install

**Reply by DeveloperKev** [reply]
i want to install gcc compiler from this binary file:
`https://ftp.gnu.org/gnu/gcc/gcc-12.2.0/gcc-12.2.0.tar.xz`
on m1 mac. where should i place the files?

**Reply by SMillerDev** [reply]
https://gcc.gnu.org/install/ documents how to install GCC.

P.S. since this has strayed far beyond Homebrew I'm going to stop answering this.

**Reply by DeveloperKev** [reply]
sure.
but where are the pre-built binaries of gcc in Homebrew? as suggested by https://gcc.gnu.org/install/binaries.html

**Reply by SMillerDev** [reply]
`brew install gcc` after you install Homebrew

**Reply by DeveloperKev** [reply]
> `brew install gcc` after you install Homebrew

does brew build `gcc` when installing?
cant the binaries be hosted on github? i've less than 10gb space free on my 2018 mac.
it'd better if i can get any help.

**Reply by SMillerDev** [reply]
> does brew build gcc when installing?

That depends on your system

>  cant the binaries be hosted on github?

They are, but not in a format that makes it possible to install them outside of brew

> i've less than 10gb space free on my 2018 mac.

That's not something I can help with

**Reply by DeveloperKev** [reply]
can u put a link to the binaries
i'm on Macos monterey.

**Reply by SMillerDev** [reply]
https://github.com/Homebrew/homebrew-core/pkgs/container/core%2Fgcc%2F12

Like I said: "They are, but not in a format that makes it possible to install them outside of brew"

**Reply by DeveloperKev** [reply]
would u mind sharing the build commands for gcc that brew uses?
whether its a workflow, i have only got for intel ones

**Reply by carlocab** [reply]
The build configuration is available at https://github.com/Homebrew/homebrew-core/blob/f554ff6487482fb9673848ba038f7e89eb70ec92/Formula/gcc.rb. All build configuration for all formulae can be found at that repository.

However, I strongly advise against using our builds, particularly for formulae like GCC, as a model for your own builds (unless you really know what you're doing).

A lot of our configuration is motivated by packaging concerns that are specific to Homebrew. They also rely on some packaging that `brew` itself does behind the scenes but aren't transparent in individual formulae.

**Reply by DeveloperKev** [reply]
why dont just brew post the `gcc` build directly?
who will want to install `xcode tools` only for installing `gcc`  with 256gb storage, the os itself takes a lot.
there are some builds available but are not trustworthy enough.
i think, homebrew can do better in here.

**Reply by SMillerDev** [reply]
> why dont just brew post the gcc build directly?

Because building compilers is a lot more complicated than you realize and a lot of work goes into ensuring the packages can be used. There is no such thing as "just post the build" since package managers also do a lot of work to install the dependencies that you need for the build.

>  who will want to install xcode tools only for installing gcc with 256gb storage, the os itself takes a lot.

People who need to compile anything

>  there are some builds available but are not trustworthy enough.

That's not something brew can fix.

>  i think, homebrew can do better in here.

You can think that, but that won't change the fact that you want something that Homebrew does not and will never support.

**Reply by carlocab** [reply]
> why dont just brew post the `gcc` build directly?

I don't know what you mean by this.

> who will want to install `xcode tools` only for installing `gcc` with 256gb storage, the os itself takes a lot.

Unless you only need the GCC runtime libraries, GCC itself will not work if you do not have the macOS SDK. To get the macOS SDK, you need to install Xcode or the Xcode CLT.

**Reply by DeveloperKev** [reply]
@SMillerDev thanks

**Reply by DeveloperKev** [reply]
one thing: is `iains/gcc-12-branch` the official repo of `gcc` used by brew (for apple silicon)?

**Reply by DeveloperKev** [reply]
I respect open-source but
How are u gonna address this? basically anyone can use brew to install unknown packages without the need of permission/ required privileges.
https://www.youtube.com/watch?v=UvobNN37SJY

Also, correct me if i'm wrong, there are so many repo's which are not well reviewed.

**Reply by carlocab** [reply]
There isn't anything that needs to be addressed:
1. if someone has access to your file system, you've already been pwned (see, for example https://xkcd.com/1200/)
2. requiring `sudo` to install packages opens you up to other attack vectors
3. if you don't like Homebrew's security model, and prefer a package manager that requires you to use `sudo`, try MacPorts

**Reply by DeveloperKev** [reply]
but the `opt/` folder is already admin priviledged. what if the mischief is done here,
isnt `sudo` a measure of confirmation to be used only under a session or for a command?

what do u say about the security flaws that has been fixed lately? doesnt it mean all the prev installed brew unsafe?
i know its a package manager, and it depends entirely in the users.

**Reply by SMillerDev** [reply]
If someone has access to your account they can change your path and you might be calling `/evil/bin/sudo` instead of `/usr/bin/sudo`.

**Reply by DeveloperKev** [reply]
> If someone has access to your account they can change your path and you might be calling `/evil/bin/sudo` instead of `/usr/bin/sudo`.

well, thats a point.

> There is no such thing as "just post the build" since package managers also do a lot of work

cant we get a dmg file of `gcc` which will install the dependencies when installing rather than installing a whole bunch of unnecessary tools.

>is iains/gcc-12-branch the official repo of gcc used by brew (for apple silicon)?

ans expected.

>  many repo's are not well reviewed.

there are versions of `brew` without `sudo` , what do u think of that?
btw its nice talking to u.

**Reply by DeveloperKev** [reply]
hello?

**Reply by DeveloperKev** [reply]
> https://github.com/Homebrew/homebrew-core/pkgs/container/core%2Fgcc%2F12
> 
> Like I said: "They are, but not in a format that makes it possible to install them outside of brew"

make it useable pls.

**Comment by Janis1970** [comment]
Thanks for info man never no nothing about it👍👍👍Sent using the mobile mail appOn 2022-08-29 at 17:49, ™ wrote:
                
            
            
                
                    From: "™" ***@***.***>Date: 29 August 2022To: "Homebrew/discussions" ***@***.***>Cc: "Subscribed" ***@***.***>Subject: Re: [Homebrew/discussions] Building for M1 Macs (Discussion #3571)
                
                
I respect open-source but
How are u gonna address this? basically anyone can use brew to install unknown packages without the need of permission/ required privileges.
https://www.youtube.com/watch?v=UvobNN37SJY
Also, correct me if i'm wrong, there are so many repo's which are not well reviewed.
